File: go/linkedlist/leetcode/palindrome_linked_list.go
package main

// 回文链表

// 使用一个数组来存储
func isPalindrome1(head *ListNode) bool {
	var arr []int
	for ; head != nil; head = head.Next {
		arr = append(arr, head.Val)
	}

	if len(arr) > 1 {
		for i, j := 0, len(arr)-1; i < j; i, j = i+1, j-1 {
			if arr[i] != arr[j] {
				return false
			}
		}
	}
	return true
}

// 原地解决
func isPalindrome2(head *ListNode) bool {
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		slow, fast = slow.Next, fast.Next.Next
	}

	mid := slow
	if fast != nil {
		mid = slow.Next
	}

	var prev *ListNode
	var next *ListNode = mid
	for next != nil {
		next.Next, prev, next = prev, next, next.Next
	}

	for head != slow && prev != nil {
		if head.Val != prev.Val {
			return false
		}
		head, prev = head.Next, prev.Next
	}
	return true
}

File: python/linkedlist/leetcode/palindrome_linked_list.py

# 回文链表

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:

    # 使用一个数组存储
    def isPalindrome_1(self, head: ListNode) -> bool:
        data = []
        while head:
            data.append(head.val)
            head = head.next
        return data == data[::-1]

    # 原地遍历，快慢指针
    def isPalindrome_2(self, head: ListNode) -> bool:
        # 找到中间节点
        slow, fast = head, head
        while fast and fast.next:
            slow, fast = slow.next, fast.next.next
        
        mid = slow
        if fast:
            mid = slow.next

        # 反转mid->tail的节点
        prev, back = None, mid
        while back:
            back.next, prev, back = prev, back, back.next
        
        # 依次遍历并比较
        while prev and head != slow:
            if prev.val != head.val:
                return False
            prev, head = prev.next, head.next
        return True
